src/dataframe/dataframe.jl

```diff
@@ -388,33 +398,51 @@ end
 # df[SingleColumnIndex] = Single Item (EXPANDS TO NROW(df) if NCOL(df) > 0)
 function Base.setindex!(df::DataFrame, v, col_ind::ColumnIndex)
     if haskey(index(df), col_ind)
+        if !(typeof(col_ind) <: Union{Integer,Symbol})
+            throw(ArgumentError("Only Integer numbers accepted for column indexing"))
+        end
+        if isa(col_ind, Bool)
+            throw(ArgumentError("Bool values are not allowed for column indexing"))
+        end
         fill!(df[col_ind], v)
     else
         insert_single_column!(df, upgrade_scalar(df, v), col_ind)
     end
-    return df
+    return v
 end
 
 # df[MultiColumnIndex] = DataFrame
-function Base.setindex!(df::DataFrame, new_df::DataFrame, col_inds::AbstractVector{Bool})
+function Base.setindex!(df::DataFrame, new_df::DataFrame,
+                        col_inds::AbstractVector{<:Union{Bool,Missing}})
+    length(col_inds) != ncol(df) && throw(BoundsError(df, col_inds))
     setindex!(df, new_df, findall(col_inds))
 end
 function Base.setindex!(df::DataFrame,
                         new_df::DataFrame,
                         col_inds::AbstractVector{<:ColumnIndex})
+    if length(col_inds) != ncol(new_df)
+        Base.depwarn("Number of columns in RHS DataFrame should be equal to number of assignments.", :insert!)
+    end
+    # after deprecation replace above warning by
+    # length(col_inds) != ncol(new_df) && throw(ArgumentError("Number of columns in RHS DataFrame must be equal to number of assignments"))
```

Comment:
No initial uppercase (also elsewhere). Also it would be nice to print the actual number of columns on both sides, that makes it much easier to understand what has gone wrong.

Reply:
fixed